File: Data Structures & Algorithms/jump-game/submission-0.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        vector<int> dp(nums.size(),-1) ;
        return solve(0,nums,dp) ;
    }
    bool solve(int i,vector<int>& nums,vector<int>& dp){

        int n = nums.size() ;
        if(i==n-1)
            return true ;
        if(dp[i]!=-1)
            return dp[i] ;
        int end = min(n-1,i+nums[i]) ;
        for(int j=i+1;j<=end;++j){
            if(solve(j,nums,dp))
                return dp[i] =  true ;
        }
        return dp[i] = false ;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        suffix =[0]*(n+1)
        for i in range(n-1,-1,-1):
            suffix[i] = max(suffix[i+1],height[i])
        prefix = 0 
        ans = 0 
        for i in range(n):
            prefix = max(prefix,height[i])
            minLen = min(prefix,suffix[i])
            ans += minLen - height[i]
        
        return ans 